plugins/inputs/cloud_pubsub/pubsub.go

```diff
@@ -170,30 +177,39 @@ func (ps *PubSub) startReceiver(parentCtx context.Context) error {
 
 // onMessage handles parsing and adding a received message to the accumulator.
 func (ps *PubSub) onMessage(ctx context.Context, msg message) error {
+	defer msg.Ack()
 	if ps.MaxMessageLen > 0 && len(msg.Data()) > ps.MaxMessageLen {
-		msg.Ack()
 		return fmt.Errorf("message longer than max_message_len (%d > %d)", len(msg.Data()), ps.MaxMessageLen)
 	}
 
+	// This function is called concurrently, but the decoder cannot.
+	ps.decoderMutex.Lock()
+	b, err := ps.decoder.Decode(msg.Data())
+	ps.decoderMutex.Unlock()
+	if err != nil {
+		return fmt.Errorf("unable to decode message: %v", err)
+	}
 	var data []byte
+	if ps.ContentEncoding == "gzip" {
+		data = make([]byte, len(b))
+		copy(data, b)
+	} else {
+		data = b
+	}
```

Comment:
Uhh I think you have to do this while holding the `ps.decoderMutex` lock, as otherwise writing to `b` might be racy.

Reply:
this is not an issue